File: strategy/generator/filter/eom.py
from dataclasses import dataclass

from core.models.parameter import Parameter, StaticParameter

from .base import Filter, FilterType


@dataclass(frozen=True)
class EomFilter(Filter):
    type: FilterType = FilterType.Eom
    period: Parameter = StaticParameter(14.0)
    divisor: Parameter = StaticParameter(10000.0)

File: ta_lib/strategies/filter/src/eom.rs
use base::prelude::*;
use core::prelude::*;
use volume::eom;

const EOM_FILTER: f32 = 0.0;

pub struct EOMFilter {
    period: usize,
    divisor: f32,
}

impl EOMFilter {
    pub fn new(period: f32, divisor: f32) -> Self {
        Self {
            period: period as usize,
            divisor,
        }
    }
}

impl Filter for EOMFilter {
    fn lookback(&self) -> usize {
        self.period
    }

    fn confirm(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let eom = eom(
            &data.hl2(),
            &data.high,
            &data.low,
            &data.volume,
            self.period,
            self.divisor,
        );

        (eom.sgt(&EOM_FILTER), eom.slt(&EOM_FILTER))
    }
}
